### ISSUE TITLE
rust_ld cannot be set to LLD

### ISSUE THREAD

**Issue by alyssais** [mesonbuild/meson#12577]
**Describe the bug**
Aside from the special case of "link", when rust_ld is set Meson assumes it's a C/C++ compiler, rather than a linker. (It calls `_detect_c_or_cpp_compiler`.)

**To Reproduce**
```ShellSession
meson init -l rust
RUSTC_LD=ld.lld meson setup builddir
```

**Expected behavior**
Meson should recognise LLD as a valid Rust linker, and pass `-C linker=ld.lld` or similar.

**system parameters**
* Is this a [cross build](https://mesonbuild.com/Cross-compilation.html) or just a plain native build (for the same computer)? <ins>native</ins>
* what operating system (e.g. MacOS Catalina, Windows 10, CentOS 8.0, Ubuntu 18.04, etc.) <ins>Debian 12</ins>
* what Python version are you using e.g. 3.8.0  <ins>3.11.2</ins>
* what `meson --version` <ins>1.3.99</ins>
* what `ninja --version` if it's a Ninja build <ins>1.11.1</ins>

**Comment by kiwixz** [comment]
Did you try `RUSTC_LD=lld` ?

**Comment by v1993** [comment]
This is kind of convoluted. Basically, `rustc` wants the linker to actually be a compiler and will fail if you give it `lld.ld`; that's why meson does what it does. What you can do instead is

```sh
RUSTC='rustc -C link-arg=--ld-path=ld.lld' RUSTC_LD=clang meson setup build
```

**Comment by alyssais** [comment]
Not exactly.  rustc supports linking with LLD and in fact now does it by default on x86_64-unknown-linux-gnu on the official binaries.  The bug is in Meson.

**Comment by bonzini** [comment]
Meson currently relies on the options passed by the driver to the linker, but you should be able to work around this by setting `rustc_ld` to `gcc -fuse-ld=lld` or similar.

**Comment by v1993** [comment]
> Not exactly. rustc supports linking with LLD and in fact now does it by default on x86_64-unknown-linux-gnu on the official binaries. The bug is in Meson.

I've specifically modified meson to use `rustc -C linker=ld.lld` and that fails:

```
  = note: ld.lld: error: unable to find library -lgcc_s
          ld.lld: error: unable to find library -lutil
          ld.lld: error: unable to find library -lrt
          ld.lld: error: unable to find library -lpthread
          ld.lld: error: unable to find library -lm
          ld.lld: error: unable to find library -ldl
          ld.lld: error: unable to find library -lc
```

> [...] you should be able to work around this by setting `rustc_ld` to `gcc -fuse-ld=lld` or similar.

That appears to work at the first glance (`RUSTC_LD='clang -fuse-ld=lld'`), but linker arguments are actually nowhere to be found in the generated build commands. Meson probably should pass linker arguments to `rustc` using `-C link-arg=...`. My version seems to work correctly.

**Comment by alyssais** [comment]
> > Not exactly. rustc supports linking with LLD and in fact now does it by default on x86_64-unknown-linux-gnu on the official binaries. The bug is in Meson.
> 
> I've specifically modified meson to use `rustc -C linker=ld.lld` and that fails:
> 
> ```
>   = note: ld.lld: error: unable to find library -lgcc_s
>           ld.lld: error: unable to find library -lutil
>           ld.lld: error: unable to find library -lrt
>           ld.lld: error: unable to find library -lpthread
>           ld.lld: error: unable to find library -lm
>           ld.lld: error: unable to find library -ldl
>           ld.lld: error: unable to find library -lc
> ```

Right, it [looks like](https://blog.rust-lang.org/2024/05/17/enabling-rust-lld-on-linux/) you're actually supposed to use rust-lld, which I guess is some wrapper.

**Comment by v1993** [comment]
> > > Not exactly. rustc supports linking with LLD and in fact now does it by default on x86_64-unknown-linux-gnu on the official binaries. The bug is in Meson.
> > 
> > 
> > I've specifically modified meson to use `rustc -C linker=ld.lld` and that fails:
> > ```
> >   = note: ld.lld: error: unable to find library -lgcc_s
> >           ld.lld: error: unable to find library -lutil
> >           ld.lld: error: unable to find library -lrt
> >           ld.lld: error: unable to find library -lpthread
> >           ld.lld: error: unable to find library -lm
> >           ld.lld: error: unable to find library -ldl
> >           ld.lld: error: unable to find library -lc
> > ```
> 
> Right, it [looks like](https://blog.rust-lang.org/2024/05/17/enabling-rust-lld-on-linux/) you're actually supposed to use rust-lld, which I guess is some wrapper.

According to the article

> When CI builds LLVM to use in the compiler, it also builds the linker and packages it. It's referred to as `rust-lld` to avoid colliding with any `lld` already installed on the user's machine.

In other words, `rust-lld` is same as `lld`, just renamed in certain distribution methods to avoid name collisions. If it was a wrapper then we'd probably still want to avoid using it to allow alternative linkers such as mold and wild.

**Comment by v1993** [comment]
@bonzini You version could be made to work by fixing `detect_rust_compiler` to handle link args in the `else` branch like it is already done if there's no override; replace

https://github.com/mesonbuild/meson/blob/74395f67e702848d5148c165ce8dbf442fc18636/mesonbuild/compilers/detect.py#L1172-L1178

with

```python
                cc = _detect_c_or_cpp_compiler(env, 'c', for_machine, override_compiler=override)
                linker = cc.linker
                exelist = cc.linker.exelist.copy()

                # Of course, we're not going to use any of that, we just
                # need it to get the proper arguments to pass to rustc
                if exelist[0].endswith('ccache'):
                    del exelist[0]

                c = exelist.pop(0)
                compiler.extend(cls.use_linker_args(c, ''))
                for l in exelist:
                    compiler.extend(['-C', f'link-arg={l}'])
```

However, it looks like that function should generally be refactored to avoid logic duplication between branches (already the case with ccache detection logic), which I'm not quite up to.
